crates/sui-oracle/src/lib.rs

```diff
@@ -0,0 +1,353 @@
+    metrics: Arc<OracleMetrics>,
+}
+
+impl DataProvider {
+    pub async fn run(&self) {
+        info!(
+            feed_name = self.feed_name,
+            source_name = self.source_name,
+            "Starting data provider."
+        );
+        let mut interval = tokio::time::interval(self.upload_feed.submission_interval);
+        interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Skip);
+
+        loop {
+            interval.tick().await;
+            self.run_once().await;
+        }
+    }
+
+    async fn run_once(&self) {
+        info!(
+            feed_name = self.feed_name,
+            source_name = self.source_name,
+            "Running data provider once."
+        );
+        let value = self.retrieve_from_data_source().await;
+        if value.is_err() {
+            error!(
+                feed_name = self.feed_name,
+                source_name = self.source_name,
+                "Failed to retrieve data from data source: {:?}",
+                value
+            );
+            self.metrics
+                .data_source_errors
+                .with_label_values(&[&self.feed_name, &self.source_name])
+                .inc();
+            return;
+        }
+
+        self.metrics
+            .data_source_successes
+            .with_label_values(&[&self.feed_name, &self.source_name])
+            .inc();
+
+        let value = value.unwrap();
+        match self.submit(value).await {
+            Ok(effects) => {
+                info!(
+                    feed_name = self.feed_name,
+                    source_name = self.source_name,
+                    "Submitted value: {value}."
+                );
+                self.metrics
+                    .uploaded_values
+                    .with_label_values(&[&self.feed_name])
+                    .observe((value * METRICS_MULTIPLIER) as u64);
+                self.metrics
+                    .upload_successes
+                    .with_label_values(&[&self.feed_name, &self.source_name])
+                    .inc();
+                let gas_usage = effects.gas_cost_summary().gas_used();
+                self.metrics
+                    .gas_used
+                    .with_label_values(&[&self.feed_name])
+                    .observe(gas_usage);
+                self.metrics
+                    .total_gas_used
+                    .with_label_values(&[&self.feed_name])
+                    .inc_by(gas_usage);
+            }
+            Err(_) => {
+                error!(
+                    feed_name = self.feed_name,
+                    source_name = self.source_name,
+                    "Failed to submit value: {value}"
+                );
+                self.metrics
+                    .upload_data_errors
+                    .with_label_values(&[&self.feed_name, &self.source_name])
+                    .inc();
+            }
+        }
+    }
+
+    async fn retrieve_from_data_source(&self) -> anyhow::Result<f64> {
```

Comment:
Should we have a type that does not lose precision here, unlike 'f64'?

Reply:
Move does not support float natively. @patrickkuo  is exploring a way to make it happen.